XSPEC: add support for XSPEC 12.15.1

DougBurke opened the pull request:
# Summary

Allow Sherpa to build against XSPEC 12.15.1 and simplify the XSPEC installation steps. Add 13 new additive models: XSbfekblor, XSfekblor, XSzfekblor, XSzbfekblor, XSbfeklor, XSzbfeklor, XSrsapec, XSrsvapec, XSrsvvapec, XSrsgaussian, XSrsrnei, XSrsvrnei, XSrsvvrnei. Fix #2392.

# Details

The documentation for the XSPEC configuration has (hopefully) been improved, and now mentions the (relatively new) HEASARC xspec builds. It turns out that with recent XSPEC builds we do not need to be as explicit as we used to with the "ancillary" libraries as they are referenced via rpath/whatever the magic is. This means that we not need to include "hdsp_<version>" in the XSPEC libraries list, nor do we need to give CCFits and wcslib libraries (again with versions). So the documenation has had the "per-XSPEC-version" list of options removed. I have also changed the XSPEC build step to default to $HEADAS/include and $HEADAS/lib for the XSPEC paths [*] and then allowed environment variables to be used for any of the paths **but only for** the XSPEC options (i.e. not FFTW or region or ...).

[*] Of cource, the CXC-provided `xspec-modelsonly` packages actually have $HEADAS pointing to somewhere different, and you actually want either $HEADAS/../lib or $CONDA_PREFIX/lib instead. Which means that the `scripts/use_ciao_config` and `.github/scrpts/setup_xspec.sh` helper scripts need minor tweaks (because the default options in `setup.cfg` have changed so the scripts don't recognize the lines to change). Other downstream code, like #1544, will also likely need to be updated.

Allows Sherpa to be built with XSPEC 12.15.1 and provide access to the 13 new additive models:

    XSbfekblor     XSfekblor    XSzfekblor    XSzbfekblor
    XSbfeklor      XSzbfeklor
    XSrsapec       XSrsvapec     XSrsvvapec
    XSrsgaussian
    XSrsrnei        XSrsvrnei       XSrsvvrnei

XSPEC 12.15.1 has added support for `"latest"` settings for the apec/atomdb, nei, and spex database versions (the names of this are not clear, e.g. apec vs atomdb, nei vs neiapec). For example:

```
% grep VERSION ~/.xspec/Xspec.init
ATOMDB_VERSION:  latest
SPEX_VERSION:  latest
NEI_VERSION: latest
```

There is now a `latest.txt` file in the `modelData` directory that can be used to find out what "latest" means, but this is not guaranteed to exist, in which case it falls back to default values. It appeas that the default depends on whether you are running XSPEC (the program) or accessing just the model library.

For now we attempt to follow the program versions (as they are newer) as the defaults (when the `latest.txt` file does not exist), but **only** when the interface is built with XSPEC 12.15.1 (or later), as it's not clear what to do with older XSPEC libraries. One option would be to just error out if `latest` is set but there's no `latest.txt` file: I decided against this as it's possible to get a 12.15.1 `Xspec.init` file, which has `latest` values in it, and then switch to use Sherpa (e.g. via CIAO) and it would be annoying for it to then fall over (although there is a strong argument to be said it would make things more obvious about what is going on, but I don't think most users would agree with this choice). 

The `xset` database now gets three values automatically associated with it. This is not technically a breaking change but it can be surprising (e.g. the `save_all` output will now include these three settings if XSPEC is installed/used):

```
>>> from sherpa.astro import xspec
>>> xspec.get_xsversion()
'12.15.1'
>>> xspec.get_xsxset()
{'APECROOT': '3.1.3', 'NEIAPECROOT': '3.1.3', 'SPEXROOT': '3.0.8'}
```

# Notes

I believe there are some issues with the XSPEC 12.15.1 release but I have not reported them yet as HEASARC is closed down. Problems include

- XSPEC initalization defaults to atomdb/nei/spex versions of 3.13/3.13/3.08 but the FORTRAN `FNINIT` code we have to use uses 3.12/3.12/3.08 (at least that is what it seems like)
- however, the $HEADAS/spectral/modelData/latest.txt uses "3.0.8" for the SPEX version and this will win out if you have a "new" ~/.xspec/Xspec.init file with "SPEX_VERSION: latest" in it. I believe this is wrong and it should be "3.08".

**EDIT** These have now been reported to the XSPEC team (Jan 14 2026).

Originally this supported `get_xsversion("atomdb")` and `set_xsversion("atomdb", "4.0.1")` but I have removed this because

a) too many ways to set the same information (e.g. `set_xsxset("APECROOT", "4.0.1")`) makes it confusing
b) if you use a pre-XSPEC 12.15.1 library then the different ways to set this information can lead to different behavior (which means we probably don't want to do this until we default to 12.15.1, if ever)

# Examples

XSPEC 12.15.1 installed via HEASARC conda packages, including the data

```
% grep VERSION ~/.xspec/Xspec.init
ATOMDB_VERSION:  latest
SPEX_VERSION:  latest
NEI_VERSION: latest
% cat $HEADAS/spectral/modelData/latest.txt 
ATOMDB_VERSION: 3.1.3
SPEX_VERSION: 3.0.8
NEI_VERSION: 3.1.3
```

```
>>> from sherpa.astro import xspec
>>> xspec.get_xsversion()
'12.15.1'
>>> xspec.get_xsxset()
{'APECROOT': '3.1.3', 'NEIAPECROOT': '3.1.3', 'SPEXROOT': '3.0.8'}
```

We can see an apec model and some of the files it loads by bumping the `chatter` setting:

```
>>> xspec.set_xschatter(50)
>>> mdl = xspec.XSapec()
>>> mdl([0.1, 0.2, 0.3], [0.2, 0.3, 0.4])
calcMultiTempPlasma : plasmaType = 6

calcMultiTempPlasma : plasmaType = 6
calcMultiTempPlasma : TraceAbund = 1

Reading APEC data from 3.1.3

Reading continuum   data from /home/dburke/micromamba/envs/xspec/heasoft/spectral/modelData/apec_v3.1.3_coco.fits
Reading line        data from /home/dburke/micromamba/envs/xspec/heasoft/spectral/modelData/apec_v3.1.3_line.fits
Reading ion balance data from /home/dburke/micromamba/envs/xspec/heasoft/spectral/modelData/apec_v3.1.0_ionbal.fits

Reading continuum data from /home/dburke/micromamba/envs/xspec/heasoft/spectral/modelData/apec_v3.1.3_coco.fits
Reading line data from /home/dburke/micromamba/envs/xspec/heasoft/spectral/modelData/apec_v3.1.3_line.fits
Reading ion balance data from /home/dburke/micromamba/envs/xspec/heasoft/spectral/modelData/apec_v3.1.0_ionbal.fits
Not applying thermal broadening
Velocity broadening : 0
Linear interpolation on tabulated temperatures
Not multi-threading over temperatures
Including electron-electron bremsstrahlung
Not doing electron density correction on normalization
Column (rsColumn) for resonance scattering 0

For continuum interpolating between 0.966887 and 1.02418 for 1
with coefficients 0.422019 and 0.577981

For lines interpolating between 0.966887 and 1.02418 for 1
with coefficients 0.422019 and 0.577981

Reading data for temperature 122 0.966887 0.966882

Reading data for temperature 122 0.966887 0.966882
Read 30417 element lines and 30  continuum points for 30 unique elements.

Reading data for temperature 123 1.02418 1.02417

Reading data for temperature 123 1.02418 1.02417
Read 28818 element lines and 30  continuum points for 30 unique elements.

array([2.01612911, 0.30307714, 0.2181515 ])
```

DougBurke committed:
XSPEC: add initial support for XSPEC 12.15.1

Allow Sherpa to build against XSPEC 12.15.1.

This provides low-level access to the new models but it does not
add the new models, which are:

  XSbfekblor     XSzbfekblor
  XSbfeklor      XSzbfeklor
  XSfekblor      XSzfekblor
  XSrsapec
  XSrsgaussian
  XSrsrnei
  XSrsvapec
  XSrsvrnei
  XSrsvvapec
  XSrsvvrnei

XSPEC 12.15.1 appears to have added support for specifying the
ATOMDB/NEI/SPEX versions to "latest" (case insensitive).

The implementation appears to differ depending on whether you
are using XSPEC or the XSPEC model library (this is presumably a
bug).

Of course, for Sherpa - which supports multiple XSPEC versions -
we have to decide

- when to worry about "latest" settings,
- and what to use for the defaults.

Since users can end up with "latest" values even when building
Sherpa with "old" XSPEC versions - as the value can be read from
~/.xspec/Xspec.init - we decide to always try to convert "latest",
and use the HEADAS 6.36/XSPEC 12.15.1 values for atomdb, nei, and
spex as the default (spex only set when XSPEC >= 12.15.0).

It looks like the APECROOT/NEIAPECROOT/SPEXROOT string-database
settings now interact with the atomdbVersion/neiVersion/spexVersion
settings (i.e. changing one changes the other). This makes our
life easier as there is less need to try and synchronize settings,
at least if you have XSPEC 12.15.1.

There are hidden changes to how the "string database" as handled by
set_xsxset/get_xsxset behave: 12.15.1 now appears to always add
several keywords (APECROOT, NEIAPECROOT, and SPEXROOT) which ends up
causing problems for our save_all/serialization tests. So the tests
have been re-written to just check that the values are set, but
not care about the actual value (via the "# doctest: +ELLIPSIS" option
added here).

As part of this work, update the installation documentation for
configuring XSPEC:

- change the labeling of the per-version guides from numeric to
  unlabeled (to simplify updates when adding new or removing old
  versions)

- add a note that if XSPEC was installed via conda then the
  settings are (likely) simpler.

DougBurke committed:
XSPEC: adding new model classes from 12.15.1

The 13 new classes (all additive) are

  XSbfekblor     XSzbfekblor
  XSbfeklor      XSzbfeklor
  XSfekblor      XSzfekblor
  XSrsapec       XSrsvapec        XSrsvvapec
  XSrsgaussian
  XSrsrnei       XSrsvrnei        XSrsvvrnei

DougBurke commented:
Rebased to address conflicts.

DougBurke commented:
Rebased to pick up some changes (with some commits having moved to other PRs).

DougBurke commented:
Rebased on top of #2444 and rework the ordering (since handling of "latest" needs to be done as soon as Sherpa is built against 12.15.1).

DougBurke commented:
XSPEC provides a number of ways of setting values. For the ATOMDB/... values this PR supports set_xsxset and set_xsversion. Perhaps we just remove the set_xsversion as (at least for 12.15.1) it looks like they all end up changing the same internal settings (it wasn't clear in earlier versions).

**EDITED TO ADD** I have now removed these additions as they didn't work well with a pre-12.15.1 system and there are too many ways to set the same information.

DougBurke committed:
XSPEC: update the documentation when building XSPEC

There are two documentation changes:

1. the setup.cfg defaults had not been updated since XSPEC 12.12.0
   and as this version is no-longer supported it makes sense to
   update it. I chose the latest version for the defaults, as that's
   most useful at this time (i.e. until there's another HEASOFT
   release).

   I also updated the text to be (hopefully) more coherent about
   what needs to be set.

2. the XSPEC installation section of the documentation included
   the so-called defaults of the setup.cfg settings, but they weren't
   the actual defaults. So I removed them as I didn't feel it actually
   helped the user (as we mention the settings in the per-version
   sections).

These changes may force down-stream code (e.g. code that updates
the setup.cfg file to match a particular environment) to change.

DougBurke committed:
XSPEC: default to $HEADAS for XSPEC lib/include directories

This also allows the user to set the path as $HEADAS/lib or
$HEADAS/include rather than manually having to expand the
variables. The build could fail if $HEADAS is not set but it seems
excessive (it is only needed to run the tests or use Sherpa, not to
build).

The installation documentation has been greatly simplified because it
should now be "simple" to build - i.e. there should be no need to
include the per-HEADAS-release outputs (thanks to modern XSPEC builds
using rpath to reference the needed libraries).

There are some changes to the "helper" code used to change the
configuration file (setup.cfg) to match the environment, and other
downstream tools may also need to be updated.

DougBurke commented:
I did not mean to assign this to @hamogu and the GH web interface is just playing up too much recently ....

DougBurke commented:
@nplee - can you check this out (the documentation is meant to suggest how to use the HEASARC XSPEC 12.15.1/HEADAS 6.36 conda build if you don't have a 12.15.1 XSPEC lying around!) ?

nplee commented:
@DougBurke: I was able to build Sherpa into Conda using both a standalone XSpec 12.15.1 build and the HEASARC's distribution of Conda XSpec 12.15.1.  I just needed to set to use the right `$HEADAS/lib` and `$HEADAS/include` paths in the `setup.cfg` file and use gcc/g++ v14, using v15 resulted in failures.  So long as I used the correct `Xspec.init` file, all the pytests ran to completion.

hamogu commented on sherpa/astro/ui/tests/test_serialize.py:
```diff
@@ -2542,6 +2544,33 @@ def mymodel_func(pars, x, xhi=None):
+    xsver = tuple([int(t) for t in toks] + [patch])
+
+    # Add in the XSET settings, but only for XSPEC 12.15.1 and
+    # later. Since the versions can be uptated, take advantage of the
```
```suggestion
    # later. Since the versions can be updated, take advantage of the
```

hamogu reviewed (approved):
Nick tried is out and I know @wmclaugh will look at the C++ code. The Python side and the docs look good to me. 
It's a constant churn to keep up with XSPEC (and to support multiple versions). Not for this PR, but one thought for the future would be to decide if it's worth the effort to keep supporting multiple versions of XSPEC. For example, if we supported the 12.15.1 only and dropped all older ones, then we wouldn't need to worry about edge cases where "latest" exists, but a user has built agains an older XSPEC version etc.

DougBurke commented:
So, as CIAO doesn't guarantee to use the latest version we can't just drop all-but-the-latest. I do think we want to do a clean up and drop older support (e.g. 12.13.x). It's not clear to me what the XSPEC version scheme means (e.g. would we want to support all 12.14.x or may be just 12.1.41) but again for a different PR (and once CIAO updates it's version).

DougBurke committed:
Update sherpa/astro/ui/tests/test_serialize.py

Co-authored-by: Hans Moritz Günther <moritz.guenther@gmx.de>

DougBurke commented:
The tests fail because of DS9 changes; it would need #2453 (or equivalent) to pass.

DougBurke commented:
Rebased as the build should be okay now ...

wmclaugh commented on docs/install.rst:
```diff
@@ -218,113 +218,64 @@ XSPEC
+"""""
+
+Modern XSPEC builds use run-time search paths to reference most of the
+libraries it uses, so the only settings that should be necessary are
```
```suggestion
libraries they uses, so the only settings that should be necessary are
```

DougBurke committed:
XSPEC: stop the doc tests running in the XSPEC module

This commit has two changes

1. stop the doctests from running in this file

   The sherpa/astro/xspec module is currently black-listed in the
   pytest.ini file, but you can still run these tests with a command
   like "pytest sherpa/astro/xspec". The problem is that the examples
   change the state of the XSPEC library in such a way that can
   cause models to segfault, since the models tend not to be written
   assuming adversarial input. There's also the fact that it can be
   hard to test things like the module version string, or some default
   settings, as they change over XSPEC releases.

   So we just drop the module from the doctest tests with the
   __doctest_skip__ line.

2. Saying that, several of the examples have been "improved" so that
   they now pass. These changes include

   - fixing a typo in a symbol name
   - using ellipsis ("...") to avoid checking string values that
     depend on where the test is run
   - update the xsxsect default value as that has been changed for
     a while
   - deal with changes to the "state/database" values caused by
     XSPEC 12.15.1 (we can not assume that get_xsxset() is the empty
     dictionary any more)

DougBurke committed:
Apply suggestions from code review

Co-authored-by: wmclaugh <wmclaugh@cfa.harvard.edu>

DougBurke commented:
I expect the tests to fail as I haven't rebased it to address some recent changes unrelated to this PR.

DougBurke commented:
Closed by #2484 - this logic is now in #2492
